### ISSUE TITLE
git diff after fixing conflict from `git cherry-pick`

### ISSUE THREAD

**Issue by stevemao** [so-fancy/diff-so-fancy#196]
I did `git cherry-pick` and there is a conflict. after fixing the conflict, I did a `git diff` and see there's some extra `+` and `-`:

<img width="828" alt="screen shot 2016-08-25 at 9 13 51 am" src="https://cloud.githubusercontent.com/assets/6316590/17951273/48f002e6-6aa4-11e6-8d6e-eda13191c675.png">

Z2l0IGRpZmYKZGlmZiAtLWNjIHNyYy9lZGl0L3NlbGVjdGlvbi5qcwppbmRleCA5ODdhMzBjLDNhM2QyMTguLjAwMDAwMDAKLS0tIHNyYy9lZGl0L3NlbGVjdGlvbi5qcworKysgc3JjL2VkaXQvc2VsZWN0aW9uLmpzCkBAQCAtMzE0LDEyIC0zMjAsOCArMzIwLDEyIEBAQCBleHBvcnRzLmlzQ29sbGFwc2VkID0gaXNDb2xsYXBzZWQKICBmdW5jdGlvbiBzZWxlY3Rpb25Gcm9tRE9NKHBtLCBvbGRIZWFkKSB7CiAgICBsZXQgc2VsID0gcG0ucm9vdC5nZXRTZWxlY3Rpb24oKQogICAgY29uc3QgZG9jID0gcG0uZG9jCiAtICBsZXQgYW5jaG9yID0gcG9zRnJvbURPTShzZWwuYW5jaG9yTm9kZSwgc2VsLmFuY2hvck9mZnNldCkKIC0gIGxldCBoZWFkID0gaXNDb2xsYXBzZWQoc2VsKSA/IGFuY2hvciA6IHBvc0Zyb21ET00oc2VsLmZvY3VzTm9kZSwgc2VsLmZvY3VzT2Zmc2V0KQogKyAgbGV0IHtwb3M6IGhlYWQsIGluTGVhZjogaGVhZExlYWZ9ID0gcG9zRnJvbURPTShzZWwuZm9jdXNOb2RlLCBzZWwuZm9jdXNPZmZzZXQpCi0gICBpZiAoaGVhZExlYWYgPiAtMSAmJiBzZWwuaXNDb2xsYXBzZWQpIHsKKysgIGlmIChoZWFkTGVhZiA+IC0xICYmIGlzQ29sbGFwc2VkKHNlbCkpIHsKICsgICAgbGV0ICRsZWFmID0gZG9jLnJlc29sdmUoaGVhZExlYWYpCiArICAgIGlmICgkbGVhZi5ub2RlQWZ0ZXIudHlwZS5zZWxlY3RhYmxlKSByZXR1cm4ge3JhbmdlOiBuZXcgTm9kZVNlbGVjdGlvbigkbGVhZiksIGFkanVzdGVkOiB0cnVlfQogKyAgfQotICAgbGV0IGFuY2hvciA9IHNlbC5pc0NvbGxhcHNlZCA/IGhlYWQgOiBwb3NGcm9tRE9NKHNlbC5hbmNob3JOb2RlLCBzZWwuYW5jaG9yT2Zmc2V0KS5wb3MKKysgIGxldCBhbmNob3IgPSBpc0NvbGxhcHNlZChzZWwpID8gaGVhZCA6IHBvc0Zyb21ET00oc2VsLmFuY2hvck5vZGUsIHNlbC5hbmNob3JPZmZzZXQpLnBvcwogIAogICAgbGV0IHJhbmdlID0gZmluZFNlbGVjdGlvbk5lYXIoZG9jLnJlc29sdmUoaGVhZCksIG9sZEhlYWQgIT0gbnVsbCAmJiBvbGRIZWFkIDwgaGVhZCA/IDEgOiAtMSkKICAgIGlmIChyYW5nZSBpbnN0YW5jZW9mIFRleHRTZWxlY3Rpb24pIHsKCgoKZ2l0IGRpZmYgLS1jb2xvcgobWzE7MzNtZGlmZiAtLWNjIHNyYy9lZGl0L3NlbGVjdGlvbi5qcxtbbQobWzE7MzNtaW5kZXggOTg3YTMwYywzYTNkMjE4Li4wMDAwMDAwG1ttChtbMTszM20tLS0gc3JjL2VkaXQvc2VsZWN0aW9uLmpzG1ttChtbMTszM20rKysgc3JjL2VkaXQvc2VsZWN0aW9uLmpzG1ttChtbMTszNW1AQEAgLTMxNCwxMiAtMzIwLDggKzMyMCwxMiBAQEAbW20gG1ttZXhwb3J0cy5pc0NvbGxhcHNlZCA9IGlzQ29sbGFwc2VkG1ttCiAgZnVuY3Rpb24gc2VsZWN0aW9uRnJvbURPTShwbSwgb2xkSGVhZCkgextbbQogICAgbGV0IHNlbCA9IHBtLnJvb3QuZ2V0U2VsZWN0aW9uKCkbW20KICAgIGNvbnN0IGRvYyA9IHBtLmRvYxtbbQobWzMxbSAtICBsZXQgYW5jaG9yID0gcG9zRnJvbURPTShzZWwuYW5jaG9yTm9kZSwgc2VsLmFuY2hvck9mZnNldCkbW20KG1szMW0gLSAgbGV0IGhlYWQgPSBpc0NvbGxhcHNlZChzZWwpID8gYW5jaG9yIDogcG9zRnJvbURPTShzZWwuZm9jdXNOb2RlLCBzZWwuZm9jdXNPZmZzZXQpG1ttChtbMzJtICsgIGxldCB7cG9zOiBoZWFkLCBpbkxlYWY6IGhlYWRMZWFmfSA9IHBvc0Zyb21ET00oc2VsLmZvY3VzTm9kZSwgc2VsLmZvY3VzT2Zmc2V0KRtbbQobWzMxbS0gICBpZiAoaGVhZExlYWYgPiAtMSAmJiBzZWwuaXNDb2xsYXBzZWQpIHsbW20KG1szMm0rKyAgaWYgKGhlYWRMZWFmID4gLTEgJiYgaXNDb2xsYXBzZWQoc2VsKSkgextbbQobWzMybSArICAgIGxldCAkbGVhZiA9IGRvYy5yZXNvbHZlKGhlYWRMZWFmKRtbbQobWzMybSArICAgIGlmICgkbGVhZi5ub2RlQWZ0ZXIudHlwZS5zZWxlY3RhYmxlKSByZXR1cm4ge3JhbmdlOiBuZXcgTm9kZVNlbGVjdGlvbigkbGVhZiksIGFkanVzdGVkOiB0cnVlfRtbbQobWzMybSArICB9G1ttChtbMzFtLSAgIGxldCBhbmNob3IgPSBzZWwuaXNDb2xsYXBzZWQgPyBoZWFkIDogcG9zRnJvbURPTShzZWwuYW5jaG9yTm9kZSwgc2VsLmFuY2hvck9mZnNldCkucG9zG1ttChtbMzJtKysgIGxldCBhbmNob3IgPSBpc0NvbGxhcHNlZChzZWwpID8gaGVhZCA6IHBvc0Zyb21ET00oc2VsLmFuY2hvck5vZGUsIHNlbC5hbmNob3JPZmZzZXQpLnBvcxtbbQogIBtbbQogICAgbGV0IHJhbmdlID0gZmluZFNlbGVjdGlvbk5lYXIoZG9jLnJlc29sdmUoaGVhZCksIG9sZEhlYWQgIT0gbnVsbCAmJiBvbGRIZWFkIDwgaGVhZCA/IDEgOiAtMSkbW20KICAgIGlmIChyYW5nZSBpbnN0YW5jZW9mIFRleHRTZWxlY3Rpb24pIHsbW20KZ2l0IGNvbmZpZyBwYWdlci5kaWZmCmdpdCBjb25maWcgcGFnZXIuc2hvdwo=

**Comment by stevemao** [comment]
[diff.txt](https://github.com/so-fancy/diff-so-fancy/files/435843/diff.txt)

**Comment by scottchiefbaker** [comment]
This is a 3-way merge... I suspect this is the problem as they're a little more hairy to deal with.

**Comment by stevemao** [comment]
Thought 3-way merge should be already fixed?

**Comment by scottchiefbaker** [comment]
Ok I spent way too long on this, but the issue is diff-highlight. Specifically a couple of the lines have ansi character changes INSIDE of the leading header. Simplified it's like this:

    [red]-[darkred]-   if (headLeaf > -1 && sel.isCollapsed) {
    [green][dark_green]++  if (headLeaf > -1 && isCollapsed(sel)) {

Note the two leading "-" are separated by a color. d-s-f is expect the prefix to be " ", "+", or "-", the mixed in ansi color is causing grief. The line in question is here:

https://github.com/so-fancy/diff-so-fancy/blob/master/libexec/diff-so-fancy.pl#L344

I don't have an easy fix at this time.

**Comment by scottchiefbaker** [comment]
I updated `ansi_reveal` which will help illustrate this problem.

```
echo "G1sxOzMzbWRpZmYgLS1jYyBzcmMvZWRpdC9zZWxlY3Rpb24uanMbW20KG1sxOzMzbWluZGV4IDk4
N2EzMGMsM2EzZDIxOC4uMDAwMDAwMBtbbQobWzE7MzNtLS0tIHNyYy9lZGl0L3NlbGVjdGlvbi5q
cxtbbQobWzE7MzNtKysrIHNyYy9lZGl0L3NlbGVjdGlvbi5qcxtbbQobWzE7MzVtQEBAIC0zMTQs
MTIgLTMyMCw4ICszMjAsMTIgQEBAG1ttIBtbbWV4cG9ydHMuaXNDb2xsYXBzZWQgPSBpc0NvbGxh
cHNlZBtbbQogIGZ1bmN0aW9uIHNlbGVjdGlvbkZyb21ET00ocG0sIG9sZEhlYWQpIHsbW20KG1sz
MW0tICAgaWYgKGhlYWRMZWFmID4gLTEgJiYgc2VsLmlzQ29sbGFwc2VkKSB7G1ttChtbMzJtKysg
IGlmIChoZWFkTGVhZiA+IC0xICYmIGlzQ29sbGFwc2VkKHNlbCkpIHsbW20KG1szMm0gKyAgICBs
ZXQgJGxlYWYgPSBkb2MucmVzb2x2ZShoZWFkTGVhZikbW20K" | base64 -d | diff-highlight | third_party/ansi-reveal/ansi-reveal.pl
```

Note the ANSI mixed in with the - and + line headings

**Comment by stevemao** [comment]
Hey @scottchiefbaker , thanks so much for the time on this!

> I don't have an easy fix at this time.

I assume you mean to patch diff-so-fancy? We can definitely fix diff-highlight if it's the right thing to do. You could [report an issue or send a PR to the main git repo](https://github.com/git/git/blob/master/Documentation/SubmittingPatches).  cc @paulirish @peff

**Comment by peff** [comment]
I don't think `diff-highlight` knows about combined diffs (i.e., 3-way or greater) at all. So I suspect it treats:

```
- foo
++bar
```

as a hunk which turned ` foo` into `+bar`. It probably needs to recognize the number of `@` characters in the hunk header lines, and then soak up that many `+`, `-`, or ` ` characters before comparing lines. Or alternatively, just recognize it's a 3-way diff and stop highlighting entirely. I wouldn't be surprised if there are lots of corner cases it does badly with, as there are not just two sides to the diff anymore.

I think it probably doesn't come up in practice much because most combined-diff hunks end up with unmatched numbers of lines anyway.

**Comment by scottchiefbaker** [comment]
I guess you could consider this a diff-highlight bug. I don't expect dh to understand three-way diffs though, maybe it should? This bug is in the logic we use to remove the leading +/-

https://github.com/so-fancy/diff-so-fancy/blob/master/libexec/diff-so-fancy.pl#L344

We need to find a way to:

1. Store the original line color
2. Remove X chars of "+", "-", " " all intermixed with a possible ansi sequence
3. Maintain the internal color state between said +/-s

Not impossible, but it's a lot more complex than the logic we had before. I don't have a good/easy solution for it yet. Honestly, this is a pretty corner case (ansi mixed in the +/-), so I'm not devoting a TON of brain power to it.

**Comment by scottchiefbaker** [comment]
The more I think about this, maybe this is an issue for diff-highlight to address. It really shouldn't be styling the line indicators, as they're not really part of the context of the change. This is only an issue with diff-highlight on three way merges.

**Comment by peff** [comment]
Yeah, I agree that `diff-highlight` should ignore combined diffs. I probably won't get to it soon, so if anybody wants to try their hand at a patch, feel free.

**Comment by peff** [comment]
I just posted a fix for `diff-highlight` to the [git list](http://public-inbox.org/git/20160831050538.53fki52cvwlmkd6y@sigill.intra.peff.net/), which simply disables highlighting of combined diffs. If you want to try it out, note that it's built on top of some other patches in the thread.  The simplest thing is probably to grab the "jk/diff-highlight-potpourri" branch from https://github.com/peff/git.

**Comment by scottchiefbaker** [comment]
I can confirm that @peff's fix corrects the above mentioned issue. It lacks the contextual highlight of `diff-highlight` but that's to be expect on three-way merges moving forward. I suppose the "fix" for d-s-f is to put this new `diff-highlight` in the third-party directory?

**Comment by scottchiefbaker** [comment]
@peff did you delete the `jk/diff-highlight-potpourri` branch? I don't see it any more.

Once I can find the newest patched `diff-highlight` I'll put that in the official d-s-f repo and we should be able to close this issue.

**Comment by peff** [comment]
@scottchiefbaker Yeah, sorry, it is merged in upstream git's `master`, and I generally delete my topic branches after that. The simplest thing is to just grab what's in the upstream master, I think.

**Comment by scottchiefbaker** [comment]
@stevemao this should be fixed as of 9e99f91f9bfb02dddce6437d5143df135d2742fb. Can you confirm that it's fixed?

**Comment by stevemao** [comment]
I confirm that it fixes my issue. Thanks guys :)
